cmd/juju/commands/bootstrap.go

```diff
@@ -832,6 +583,327 @@ See `[1:] + "`juju kill-controller`" + `.`)
+		}
+		if !cons.HasArch() {
+			return true, errors.Errorf("--bootstrap-image must be used with --bootstrap-constraints, specifying architecture")
+		}
+	}
+	if c.showClouds {
+		return true, printClouds(ctx, c.ClientStore())
+	}
+	if c.showRegionsForCloud != "" {
+		return true, printCloudRegions(ctx, c.showRegionsForCloud)
+	}
+
+	return false, nil
+}
+
+func (c *bootstrapCommand) getCloud(ctx *cmd.Context) (*jujucloud.Cloud, error) {
+	bootstrapFuncs := getBootstrapFuncs()
+
+	// Get the cloud definition identified by c.Cloud. If c.Cloud does not
+	// identify a cloud in clouds.yaml, but is the name of a provider, and
+	// that provider implements environs.CloudRegionDetector, we'll
+	// synthesise a Cloud structure with the detected regions and no auth-
+	// types.
+	cloud, err := jujucloud.CloudByName(c.Cloud)
+	if errors.IsNotFound(err) {
+		ctx.Verbosef("cloud %q not found, trying as a provider name", c.Cloud)
+		provider, err := environs.Provider(c.Cloud)
+		if errors.IsNotFound(err) {
+			return nil, errors.NewNotFound(nil, fmt.Sprintf("unknown cloud %q, please try %q", c.Cloud, "juju update-clouds"))
+		} else if err != nil {
+			return nil, errors.Trace(err)
+		}
+		detector, ok := bootstrapFuncs.CloudRegionDetector(provider)
+		if !ok {
+			ctx.Verbosef(
+				"provider %q does not support detecting regions",
+				c.Cloud,
+			)
+			return nil, errors.NewNotFound(nil, fmt.Sprintf("unknown cloud %q, please try %q", c.Cloud, "juju update-clouds"))
+		}
+		var cloudEndpoint string
+		regions, err := detector.DetectRegions()
+		if errors.IsNotFound(err) {
+			// It's not an error to have no regions. If the
+			// provider does not support regions, then we
+			// reinterpret the supplied region name as the
+			// cloud's endpoint. This enables the user to
+			// supply, for example, maas/<IP> or manual/<IP>.
+			if c.Region != "" {
+				ctx.Verbosef("interpreting %q as the cloud endpoint", c.Region)
+				cloudEndpoint = c.Region
+				c.Region = ""
+			}
+		} else if err != nil {
+			return nil, errors.Annotatef(err,
+				"detecting regions for %q cloud provider",
+				c.Cloud,
+			)
+		}
+		schemas := provider.CredentialSchemas()
+		authTypes := make([]jujucloud.AuthType, 0, len(schemas))
+		for authType := range schemas {
+			authTypes = append(authTypes, authType)
+		}
+		// Since we are iterating over a map, lets sort the authTypes so
+		// they are always in a consistent order.
+		sort.Sort(jujucloud.AuthTypes(authTypes))
+		cloud = &jujucloud.Cloud{
+			Type:      c.Cloud,
+			AuthTypes: authTypes,
+			Endpoint:  cloudEndpoint,
+			Regions:   regions,
+		}
+	} else if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if err := checkProviderType(cloud.Type); errors.IsNotFound(err) {
+		// This error will get handled later.
+	} else if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return cloud, nil
+}
+
+// Get the credentials and region name.
+func (c *bootstrapCommand) getCredentialsAndRegionName(
```

Comment:
As with above, this returns lots of variables to keep the diff clean. Another one for a followup?